File: web/api/controllers/translationController.go
package controllers

import (
	"api/models"
	"api/services"
	"encoding/json"
	"fmt"
	"net/http"
)

// HandleTranslationRequest handles both PDF and text input for translation
func HandleTranslationRequest(w http.ResponseWriter, r *http.Request) {
	// Decode the incoming request
	request, err := decodeTranslationRequest(r)
	if err != nil {
		http.Error(w, fmt.Sprintf("Invalid request: %s", err), http.StatusBadRequest)
		return
	}

	// Initialize services
	pdfService := services.PDFService{}
	translationService := services.TranslationService{}

	// Extract text from the PDF or use the provided text
	textToTranslate, err := extractTextForTranslation(request, pdfService)
	if err != nil {
		http.Error(w, fmt.Sprintf("Error extracting text: %s", err), http.StatusInternalServerError)
		return
	}

	// Translate the extracted or provided text
	translationResponse, err := translateText(textToTranslate, request.SourceLanguage, request.TargetLanguages, translationService)
	if err != nil {
		http.Error(w, fmt.Sprintf("Error translating text: %s", err), http.StatusInternalServerError)
		return
	}

	// Send the translated text back to the client
	sendTranslationResponse(w, translationResponse)
}

// decodeTranslationRequest decodes the incoming request body into a TranslationRequest model
func decodeTranslationRequest(r *http.Request) (models.TranslationRequest, error) {
	var request models.TranslationRequest
	if r.Header.Get("Content-Type") != "application/json" {
		return request, fmt.Errorf("expected application/json")
	}
	err := json.NewDecoder(r.Body).Decode(&request)
	return request, err
}

// extractTextForTranslation extracts the text from either a PDF or directly from the text provided
func extractTextForTranslation(request models.TranslationRequest, pdfService services.PDFService) (string, error) {
	if request.PDFFile != nil {
		// Extract text from the PDF
		textToTranslate, err := pdfService.ExtractTextFromPDF(string(request.PDFFile))
		if err != nil {
			return "", fmt.Errorf("failed to extract text from PDF: %w", err)
		}
		return textToTranslate, nil
	}

	// If it's a text request, use the text directly
	return request.SourceText, nil
}

// translateText uses the TranslationService to translate the text into specified target languages
func translateText(textToTranslate, sourceLanguage string, targetLanguages []string, translationService services.TranslationService) (models.Translation, error) {
	translationResponse, err := translationService.TranslateText(textToTranslate, sourceLanguage, targetLanguages)
	if err != nil {
		return models.Translation{}, fmt.Errorf("translation failed: %w", err)
	}
	return translationResponse, nil
}

// sendTranslationResponse sends the translation response back to the client
func sendTranslationResponse(w http.ResponseWriter, response models.Translation) {
	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(response); err != nil {
		http.Error(w, "Failed to send response", http.StatusInternalServerError)
	}
}

File: backend/controllers/translationController.py
from fastapi import Request, HTTPException
from models.requests import TranslationRequest
from services.pdf_service import PDFService
from services.docx_service import DOCXService
from services.translation_service import TranslationService

async def handle_translation_request(request: Request):
    try:
        # Decode request
        request_data = await request.json()
        translation_request = TranslationRequest(**request_data)
        
        # Initialize services
        pdf_service = PDFService()
        docx_service = DOCXService()
        translation_service = TranslationService()
        
        # Extract text
        text_to_translate = await extract_text_for_translation(translation_request, pdf_service, docx_service)
        #maybe add a flag here to determine whether or not the translated text needs 
        #to be placed in a docx or pdf

        # Translate
        translation = await translation_service.translate_text(
            text_to_translate, #maybe implement chunking here? ensure the string isnt too big per translation request
            translation_request.source_language,
            translation_request.target_languages
        )
        
        return translation
        
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

async def extract_text_for_translation(request: TranslationRequest, pdf_service: PDFService, docx_service: DOCXService) -> str:
    if request.pdf_file:
        return await pdf_service.extract_text_from_pdf(request.pdf_file)
    elif request.docx_file:
        return await docx_service.extract_text_from_docx(request.docx_file)
    elif request.source_text:
        return request.source_text
    else:
        raise Exception("No input provided: must include a PDF file, DOCX file, or source_text.")